Interop/StubGenerator/src/main/kotlin/org/jetbrains/kotlin/native/interop/gen/StubIrTextEmitter.kt

```diff
@@ -0,0 +1,670 @@
+        private tailrec fun getDeclarationName(scope: KotlinScope, name: String): String =
+                scope.declareProperty(name) ?: getDeclarationName(scope, name + "_")
+
+        override fun visitConstructor(constructorStub: ConstructorStub, owner: StubContainer?) {
+            constructorStub.annotations.forEach {
+                out(renderAnnotation(it))
+            }
+            val visibility = renderVisibilityModifier(constructorStub.visibility)
+            out("${visibility}constructor(${constructorStub.parameters.joinToString { renderFunctionParameter(it) }}) {}")
+        }
+
+        override fun visitPropertyAccessor(propertyAccessor: PropertyAccessor, owner: StubContainer?) {
+
+        }
+
+        override fun visitSimpleStubContainer(simpleStubContainer: SimpleStubContainer, owner: StubContainer?) {
+            if (simpleStubContainer.meta.textAtStart.isNotEmpty()) {
+                out(simpleStubContainer.meta.textAtStart)
+            }
+            simpleStubContainer.classes.forEach {
+                out("")
+                it.accept(this, simpleStubContainer)
+            }
+            simpleStubContainer.functions.forEach {
+                out("")
+                it.accept(this, simpleStubContainer)
+            }
+            simpleStubContainer.properties.forEach {
+                out("")
+                it.accept(this, simpleStubContainer)
+            }
+            simpleStubContainer.typealiases.forEach {
+                out("")
+                it.accept(this, simpleStubContainer)
+            }
+            simpleStubContainer.simpleContainers.forEach {
+                out("")
+                it.accept(this, simpleStubContainer)
+            }
+            if (simpleStubContainer.meta.textAtEnd.isNotEmpty()) {
+                out(simpleStubContainer.meta.textAtEnd)
+            }
+        }
+    }
+
+    private fun EnumEntryStub.isMoreCanonicalThan(other: EnumEntryStub): Boolean = with(other.name.toLowerCase()) {
+        contains("min") || contains("max") ||
+                contains("first") || contains("last") ||
+                contains("begin") || contains("end")
+    }
+
+    // About method naming convention:
+    // - "emit" prefix means that method will call `out` by itself.
+    // - "render" prefix means that method returns string that should be emitted by caller.
+    private fun emitEnumBody(enum: ClassStub.Enum) {
+        val canonicalsByValue = enum.entries
+                .groupingBy { it.constant.value }
+                .reduce { _, accumulator, element ->
+                    if (element.isMoreCanonicalThan(accumulator)) {
+                        element
+                    } else {
+                        accumulator
+                    }
+                }
+
+        val (canonicalConstants, aliasConstants) = enum.entries.partition { canonicalsByValue[it.constant.value] == it }
```

Comment:
This logic is not a part of text emitter, it should be applied during stub building.